components/omega/doc/design/Halo.md

```diff
@@ -0,0 +1,260 @@
+
+          std::vector<I4> indices[haloWidth];   ///< list of local indices 
+
+       friend class Neighbor;
+       friend class Halo;
+    };
+
+The Neighbor class contains all the information and buffer memory needed for 
+carrying out each type of halo exchange with one neighboring rank. It contains 
+the ID of the neighboring MPI rank, an object of the ExchList class for sends 
+and receives for each mesh index space, a send and a receive buffer, MPI request 
+handles that are returned by ```MPI_Irecv``` and ```MPI_Isend``` which are 
+needed to test for completion of an individual message transfer, and boolean 
+switches to control the progress of a Halo exchange. To avoid the need for 
+separate integer and floating-point buffers, during the packing process integer 
+values can be recast as reals in a bit-preserving manner using 
+```reinterpret_cast``` and then recast as integers on the receiving end. A 
+method for determining the buffer size would be needed and would generally 
+depend on the largest field array being utilized by a run.
+
+    class Neighbor {
+
+       private:
+
+          I4 rankID;                           ///< ID of neighboring MPI rank
+          ExchList sendLists[3], recvLists[3]; ///< 0 = onCell, 1 = onEdge, 
+                                               ///< 2 = onVertex
+          std::vector<Real> sendBuffer, recvBuffer;
+          MPI_Request rReq, sReq;              ///< comm request handles
+          bool received = false;
+          bool unpacked = false;
+
+       friend class Halo;
+    };
+
+The Halo class collects all Neighbor objects needed by an MPI rank to perform 
+a full halo exchange with each of its neighbors. This class will be the user 
+interface for halo exchanges, it is a friend class to the subordinate classes 
+so that it has access to all private data.
+
+    class Halo {
+
+       private:
+
+          I4 nNghbr;                           ///< number of neighboring ranks
+          std::vector<Neighbor> neighbors;
+
+       public:
+
+          // methods
+
+    };
+
+
+### 4.2 Methods
+
+#### 4.2.1 Constructors
+
+The constructor for the Halo class will be the interface for declaring an 
+instance of the Halo class and instances of all associated member classes, 
+and depends on a Decomp object.
+
+    Halo(Decomp inDecomp);
+
+The constructors for Neighbor and ExchList will be called from within the Halo 
+constructor and will create instances of these classes based on info from the 
+Decomp object.
+
+#### 4.2.2 Field halo exchange
+
+The primary use for the Halo class will be a member function called 
+```exchangeFullFieldHalo``` which will exchange halo elements for the input 
+field with each of its neighbors across all layers of the halo. Info from the 
+machine environment is needed to carry out MPI communications. It will return 
+an integer error code to catch errors (likewise for each subprocess below).
+
+    int exchangeFullFieldHalo(Field &field, MachEnv env);
+
```

Comment:
After submitting the PR, I realized it probably would have made more sense to just save what is needed from the MachEnv in the Halo object, since it's just a couple integers(or so I thought), so I'll do that (finishing up the changes now). This did lead me to looking into the how the communicator is treated is treated in C. While in Fortran the communicator is just an integer, in C the type is MPI_Comm which is implementation dependent. I did some simple tests and found that while in MPICH the MPI_Comm type appears to just be an integer and the MPI routines accept an integer in place of the communicator, in OpenMPI and the Spectrum MPI on Summit the communicator is not interchangeable with an integer. The MachEnv class should probably be edited to store it as an MPI_Comm.

Reply:
Thanks Brian - I'll make the change in the MachEnv.